cpp/unit/producer.cpp

```diff
@@ -242,11 +239,15 @@ std::shared_ptr<TerrainObject> ObjectProducer::place(Unit *u, Terrain &terrain,
 	 * currently unit avoids water and tiles with another unit
 	 * this function should be true if pos is a valid position of the object
 	 */
-	obj->passable = [obj, &terrain, terrains](const coord::phys3 &pos) -> bool {
+	std::weak_ptr<TerrainObject> obj_ptr = u->location;
+	std::weak_ptr<Terrain> terrain_ptr = terrain;
+	u->location->passable = [obj_ptr, terrain_ptr, terrains](const coord::phys3 &pos) -> bool {
+		auto obj = obj_ptr.lock();
```

Comment:
if location is deleted, then so is this lambda, (deleting terrain implies location is deleted), so this is safe.